src/core/providers-strategy/LedgerProviderStrategy.ts

```diff
@@ -1,102 +1,154 @@
-      }
+          const data = await getLedgerProvider();
 
-      resolve(data);
-    } catch (err) {
-      if (!shouldInitiateLogin) {
-        throw err;
-      }
+          if (eventBus) {
+            eventBus.unsubscribe(
+              LedgerConnectEventsEnum.CONNECT_DEVICE,
+              onRetry
+            );
+            eventBus.unsubscribe(LedgerConnectEventsEnum.CLOSE, onCancel);
+          }
+
+          resolve(data);
+        } catch (err) {
+          if (!shouldInitiateLogin) {
+            throw err;
+          }
 
-      const { errorMessage, defaultErrorMessage } = getLedgerErrorCodes(err);
-      manager?.updateConnectScreen({
-        error: errorMessage ?? defaultErrorMessage ?? failInitializeErrorText
+          const { errorMessage, defaultErrorMessage } =
+            getLedgerErrorCodes(err);
+          manager?.updateConnectScreen({
+            error:
+              errorMessage ?? defaultErrorMessage ?? failInitializeErrorText
+          });
+
+          if (eventBus) {
+            eventBus.subscribe(LedgerConnectEventsEnum.CONNECT_DEVICE, onRetry);
+            eventBus.subscribe(LedgerConnectEventsEnum.CLOSE, onCancel);
+          }
+        }
       });
 
-      if (eventBus) {
-        eventBus.subscribe(LedgerConnectEventsEnum.CONNECT_DEVICE, onRetry);
-        eventBus.subscribe(LedgerConnectEventsEnum.CLOSE, onCancel);
-      }
+      this.config = ledgerConfig;
+      this.provider = ledgerProvider;
+      this._login = ledgerProvider.login.bind(ledgerProvider);
+      this._signTransactions =
+        ledgerProvider.signTransactions.bind(ledgerProvider);
     }
-  });
 
-  const createdProvider = provider as unknown as IProvider;
+    return this.buildProvider();
+  };
+
+  private buildProvider = async () => {
+    if (!this.provider) {
+      throw new Error(ProviderErrorsEnum.notInitialized);
+    }
 
-  const hwProviderLogin = provider.login;
+    const provider = this.provider as unknown as IProvider;
+    provider.setAccount({ address: this.address });
+    provider.signTransactions = this.signTransactions;
+    provider.login = this.login;
 
-  createdProvider.getType = () => ProviderTypeEnum.ledger;
+    await provider.init();
+    return provider;
+  };
 
-  createdProvider.signTransactions = async (
-    transactions: Transaction[]
-  ): Promise<Transaction[]> => {
-    const { ledgerProvider: customProvider } = await getLedgerProvider();
+  private createEventBus = async () => {
+    const shouldInitiateLogin = !getIsLoggedIn();
 
-    // TODO: extract to method what is below
-    const signModalElement = document.createElement(
-      'sign-transactions-modal'
-    ) as SignTransactionsModal;
+    let eventBus: IEventBus | undefined;
 
-    document.body.appendChild(signModalElement);
-    await customElements.whenDefined('sign-transactions-modal');
+    if (shouldInitiateLogin) {
+      eventBus = await getEventBus('ledger-connect-modal');
```

Comment:
there should always be a one-way flow between opening the modal and getting the event bus. I can't see the need of a split

`+` there is a different eventBus for signing transactions